src/psyclone/psyGen.py

```diff
@@ -3908,3 +3912,508 @@ def coloured_text(self):
+                                code block.
+        :rtype: :py:class:`psyclone.CodeBlock`
+        '''
+        if not statements:
+            return None
+
+        code_block = CodeBlock(statements, parent=parent)
+        parent.addchild(code_block)
+        del statements[:]
+        return code_block
+
+    # TODO remove nodes_parent argument once fparser2 AST contains
+    # parent information (fparser/#102).
+    def process_nodes(self, parent, nodes, nodes_parent):
+        '''
+        Create the PSyclone IR of the supplied list of nodes in the
+        fparser2 AST. Currently also inserts parent information back
+        into the fparser2 AST. This is a workaround until fparser2
+        itself generates and stores this information.
+
+        :param parent: Parent node in the PSyclone IR we are constructing.
+        :type parent: :py:class:`psyclone.psyGen.Node`
+        :param nodes: List of sibling nodes in fparser2 AST.
+        :type nodes: list of :py:class:`fparser.two.utils.Base`
+        :param nodes_parent: the parent of the supplied list of nodes in \
+                             the fparser2 AST.
+        :type nodes_parent: :py:class:`fparser.two.utils.Base`
+        '''
+        code_block_nodes = []
+        for child in nodes:
+            # TODO remove this line once fparser2 contains parent
+            # information (fparser/#102)
+            child._parent = nodes_parent  # Retro-fit parent info
+
+            try:
+                psy_child = self._create_child(child, parent)
+            except NotImplementedError:
+                # If child type implementation nor found add them on the
+                # ongoing code_block
+                code_block_nodes.append(child)
+            else:
+                if psy_child:  # Child could have been safely ignored.
+                    self.nodes_to_code_block(parent, code_block_nodes)
+                    # and then connect new PSyIRe child to AST
+                    parent.addchild(psy_child)
+
+        # Complete any unfinished code-block
+        self.nodes_to_code_block(parent, code_block_nodes)
+
+    def _create_child(self, child, parent=None):
+        '''
+        Create a PSyIRe node representing the the supplied fparser 2 node.
+
+        :param child: node in fparser2 AST.
+        :type child:  :py:class:`fparser.two.utils.Base`
+        :param parent: Parent node in the PSyclone IR we are constructing.
+        :type parent: :py:class:`psyclone.psyGen.Node`
+        :raises NotImplementedError: There isn't a handler for the provided \
+                child type.
+        :return: Returns the PSyIRe representation of child.
+        :rtype: :py:class:`psyclone.psyGen.Node`
+        '''
+        handler = self.handlers.get(type(child))
+        if not handler:
+            # if handler not found directly check with the base class.
+            handler = self.handlers.get(type(child).__bases__[0])
```

Comment:
I use it to provide handlers for generic classes. For instance, instead of providing an entry to the handlers map for: `Mult_Operand`, `Add_Operand`, `Level_2_Expr`, ... I just create a single one for all the BinaryOpBase subclasses.

In terms of implementation, I can't use `isinstance` because I don't check the type yet, I just get the type as a key for the handler map `get` method. I split the statement to make it more readable though.

Reply:
Thanks for explaining. Please add something along the lines of your first paragraph above to the code to explain what is happening and why. Please also change the wording about base class as (my understanding is that) the base class is not necessarily the same as the parent class and I think you mean the parent class.

Reply:
Done